### PR TITLE
an attempt to document known API problems

### PR DESCRIPTION
**Pull request by ghost**
Add notes concerning the way how we plan to tackle API.
(Ultimately, making it decent.)

### CODE REVIEW

**Comment by matejkosik** [comment]
The unfinished part is `API/PROBLEMS`. I am trying to transform my notes concerning this to a form that can be shared with everyone...

File: API/PROBLEMS/abuse/of_type_aliasing [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+but still far from ideal. We are abusing type-aliasing to avoid using records. Why?
+Why not simply define the same type as:
+
+	type definition_kind = {locality : bool;
+                                polymorphic : bool;
+                                definition_object_kind : definition_object_kind}
+
+We would not only make clear the meaning of individual fields in the type definition itself,
+but we would at the same time ensure that the meaning of the corresponding values is clear.
```
Comment (SkySkimmer) [inline API/PROBLEMS/abuse/of_type_aliasing]: For this we can define types isomorphic to bool, eg `type locality = Local | NotLocal`.

File: API/PROBLEMS/with_identifiers/strange [hunk 2]
```diff
@@ -0,0 +1,12 @@
+--------------------------------------------------------------------------------
+
+INSTANCES
+
+- Namegen.next_ident_away
+- Term.types (* plural: misleading *)
+- EConstr.types (* plural: misleading *)
+- Term.Unsafe.equal_nouniv
```
Comment (SkySkimmer) [inline API/PROBLEMS/with_identifiers/strange]: I can't find this `Term.Unsafe` module or the `equal_nouniv` function, where are they? https://github.com/coq/coq/search?q=equal_nouniv&type=Code&utf8=%E2%9C%93

Comment (matejkosik) [inline API/PROBLEMS/with_identifiers/strange]: That was a mistake. I've pushed a commit that updates that reference.

File: API/PROBLEMS/things_to_check/metavariable [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+- Why they must be part of EConstr?
+
+--------------------------------------------------------------------------------
+
+INSTANCES:
+- Metavariable.*
+- Term.mkMeta
+- Term.subst_meta
+- EConstr.pop
+- EConstr.mk_new_meta
```
Comment (Zimmi48) [inline API/PROBLEMS/things_to_check/metavariable]: Metavariables are an antique feature that Matthieu is working on getting rid of. They are currently used in the unification algorithm `w_unify`.

File: API/PROBLEMS/uncategorized [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+(things in there are not ordered randomly)
+
+It would make sense to consider extending this idea to all the modules present in the API.
+
+- make sure that it is followed by all modules that are provided by the API
+- update API/PROPERTIES accordingly
+
+--------------------------------------------------------------------------------
+
+What is the plan about the removal of API elements that are currently marked as [@@ocaml.deprecated] ?
```
Comment (Zimmi48) [inline API/PROBLEMS/uncategorized]: AFAIK these deprecation notices were added in 8.7 with the plan of getting rid of them ASAP (in 8.8).

File: API/PROBLEMS/abuse/of_type_aliasing (hunk 1 above)
Comment (matejkosik) [inline API/PROBLEMS/abuse/of_type_aliasing]: I've pushed a commit where in which I drop the example (it was not relevant in this case) and, instead, added a suggestion you mention.

File: API/PROBLEMS/with_identifiers/strange [hunk 5]
```diff
@@ -0,0 +1,11 @@
+--------------------------------------------------------------------------------
+
+INSTANCES
+
+- Namegen.next_ident_away
+- Term.types (* plural: misleading *)
+- EConstr.types (* plural: misleading *)
```
Comment (Zimmi48) [inline API/PROBLEMS/with_identifiers/strange]: That's just a typical thing when we want an identifier that is already a keyword. Some prefer `types_`. But in any case, I don't feel like this is a big issue.

**Comment by Zimmi48** [comment]
Hi Matej,
There are a lot of interesting remarks here. Not all of them are exclusive to the API though. Typically, a lot of these remarks interact with `dev/doc/style.txt` which was recently updated (#267) but could probably be updated further and eventually integrated in the upcoming `CONTRIBUTING` guide. We could e.g. define a set of prescriptive properties which shall be strictly enforced on any new code while we strive to update older code to follow the most important ones.
As for the rest of the content (the part that is API specific) I have a slight problem with its proposed location, because it is non-conventional. This would IMHO be better suited for a CEP and we could e.g. use https://github.com/coq/coq/projects/5 (possibly with a different organization than today), to track what is already done and what remains to be done.

**Comment by matejkosik** [comment]
@Zimmi48 I am not sure about the best form in which we want to maintain this information. That was one of the things that was troubling me. For the moment, consider the contents of `API/PROPBLEMS` as notes. One thing I had in mind was that I wanted to be able to:
- distinguish between:
  - a category of problems
  - and instances of this problem
- see all (known) instances of a given problem
- attach some text to a category of a problem
  (it can't just be a label)
- maintain a hierarchy of problem categories (from the most abstract ones, to the most concrete ones).
  (there are many of them)

**Comment by Zimmi48** [comment]
Well, GitHub has an API that you can query. @maximedenes has been using this for generating the CHANGES for 8.6.1 from the 8.6.1 milestone PR titles. That being said, I'm not strongly advocating for the use of GitHub projects.

I find what you want to achieve really interesting. Coq source code has already a lot of `FIXME` that we can grep but nobody is taking care of. You could indeed add some more, maybe with a more normalized content, e.g.:

    (* FIXME: use a record instead of a tuple, see recommendation XX in dev/doc/style.txt *)
    (* FIXME: use a flag for argument XX, see ... *)

And also mention these somewhere as easy projects for new contributors.

**Comment by matejkosik** [comment]
> Well, GitHub has an API that you can query.

I am aware of it but I haven't mastered it yet. I got confused at the access-control and authentication.

**Comment by herbelin** [comment]
A lot of very relevant and interesting comments. The dream of a roughly consistent roughly consensual API  satisfying-enough to be stable maybe in sight? If yes, maybe worth thinking at a "Coq 9.0" project.

**Comment by matejkosik** [comment]
@silene Thanks for the patience. I've [updated](https://github.com/coq/coq/pull/866/commits/e41f86b8e2c5adfc4f0a66b0c062d561e4d5a1e8) the files wrt. your comments.

**Comment by maximedenes** [comment]
Creating files with "*" in their names looks like a recipe for disaster. At least, it makes the Windows build fail.

**Comment by matejkosik** [comment]
I have reduced the ambitions of this PR. I think that, as a first step, it would make sense to agree upon a way how to keep track of various API problems. I have updated the branch in this respect. The commit message now says:

```
We know that Coq API is imperfect (to say the least).
It has many problems.
Some of them simpler, some of them more serious.
Some of them easier to fix, some of them harder to fix.
Some of them more and some of them less relevant (for plugin writers).

While we could simply keep a list of problems
(either in a text file, or in a bugzilla)
we would quickly realize that it is hard to avoid adding duplicate entries for the same problem(s).
We want to avoid that because that would increase the noise.

Also, if we tried simply to keep a list of problems
(either in a text file, or in a bugzilla)
it would be difficult for us to see "all problems of the same particular kind".
(if we used Jira instead of Bugzilla, then we might be able to use tags to address this problem)

Also, since there can be many distinct categories of problems
Also, it appears that it is possible to define "kinds of problems",
it is natural to try to organize them hierarchically (if possible)
instead of having a linear list of "kinds of problems".

I am not sure what tool could be collectively used to maintain such information
(about "kinds of problems" as well as their particular "instances").
Originally, I maintained this information in my personal mindmap.
That worked, but the drawback were that:
- it was not possible to share
- it was outside of the versioning system
The simplest solution I was able to find was to convert the original information
stored in the mindmap into directories and files.

Each file represents a single "kind of a problem".
The names of files are supposed to be the names of "kinds of problems".
Each file starts with a general description of a given "kind of a problem".
This is followed by a vertical line: ===================================== ... =
which is followed by a list of "instances of a given problem".
When I knew thatelipsis at the list of instances.

TODO:
- identify more "kinds of problems"
  (this may result in restructuring the hierarchy of "kinds of problems")
- identify more "instances" of the known "kinds of problems"

MOTIVATION:
- If we want to define a roadmap, we need to be aware of the problems that has to be solved
  (before we can hope to deliver an API that is not shocking).
```

**Comment by Zimmi48** [comment]
Cross-posting from the Coqdev thread:

> Let me start by saying that I share your concern on what's the right location for such information about known issues. However, I do not agree with your proposed solution.
>
> The main problem I see with this PR is that it is 2000 lines long. Most people will not read the contents then, moreover it is doomed to be largely ignored and never updated (or only by you).
A bug tracker is IMHO a much better suited place to list such problems. I understand your reticence to use Bugzilla, given the number of unsolved (and completely ignored bugs) over there, given also that some developers admittedly don't look very often at Bugzilla.
> 
> My second concern with this PR is that it is supposed to document problems of the API but most if not all of it is actually documenting much more general coding problems (I have also already said that in the PR thread).
> 
> My view on this information is that a short summary of recommended best practices should appear in Coding style guidelines that would be linked from the Contributing guide (such style guidelines already exist but are mostly ignored) and bug reports should be opened for each issue. The list of instances of the problem you give in each case is very valuable. That would make a very good checklist if only we were using GitHub issues instead of Bugzilla. Then we could check boxes one after one, when solving instances of the problem, until it disappears completely and the issue can be closed. The categorization issue you mention is IMHO much less important: in fact, I don't see much value to these categorizations being present or not (sorry about that). There are less than 50 files added by your PR, so that would correspond to less than 50 bug reports and I suppose this is bearable.
> 
> Let me add to conclude that I am currently reflecting on how we track "known" issues ("known" at least to some of us). We have a bug tracker, we write FIXME in the code, we have wishlist pages on Cocorico, we have CEPs, we have roadmaps, we had a dead TODO list in the repo until recently, and we still have TODO lists in other repositories. Moreover, many times we know about some issues but write them nowhere or only mention them in Coq-Club / Coqdev / Gitter threads or orally during a WG. My current opinion on this is that this disparity is harmful and one of its cause is that we are not really satisfied with our bug tracker (even senior developers are not satisfied with it, proof being that they are reluctant to open bug reports when they know something is wrong). One major source of dissatisfaction with our bug tracker is the inability to edit messages. If this was possible the first message could often be used to summarize the situation (when the issue is complex and the conversation too long) and even CEPs could go through issues instead of the current (non-working either) process. But I'll come back to what I discuss in this paragraph specifically in another thread.
> 
> In any case, thanks for your work on making information on these problems explicit.

**Comment by matejkosik** [comment]
@Zimmi48 I've dropped the commit which attempted to API problems. I kept the commit which removes the `API/PROPERTIES` file. This commit still make sense, I think.

**Comment by Zimmi48** [comment]
Thanks @matejkosik, although I feel like this is a very quick reaction to my message. Maybe some people would have wanted to contradict me, or at the very least have a look at what you initially suggested.
May I suggest that you push the commit attempting to define API problems again on this branch, for the record, and open a separate PR which can be merged quickly for the `API/PROPERTIES` file removal?

**Comment by matejkosik** [comment]
> May I suggest that you push the commit attempting to define API problems again on this branch, for the record, and open a separate PR which can be merged quickly for the API/PROPERTIES file removal?

I reverted the state of the branch to the original state.

**Pull request closed without merging**